File: bg_rates.pl
#! /usr/bin/perl -w
use strict;

=head1 NAME

bg_extract.pl - whee

=head1 SYNOPSIS

cp template newprog

=head1 DESCRIPTION

Extract background data from unexposed plates in the HRC-S flat field
lab data.

=head1 OPTIONS

=over 4

=item --help

Show help and exit.

=item --version

Show version and exit.

=back

=head1 AUTHOR

Pete Ratzlaff E<lt>pratzlaff@cfa.harvard.eduE<gt> February 2007

=head1 SEE ALSO

perl(1).

=cut

my $version = '0.1';

use FindBin;
use Config;
use Carp;
use MyRDB;
use PDL;
use PDL::Graphics::PGPLOT;
use Chandra::Tools::Common qw( read_bintbl_cols );
use IO::Handle;
use Data::Dumper;
use Lab;

use Getopt::Long;
my %default_opts = (
		    evtdir => $Lab::EVTDIR,
		    bgdir => $Lab::BGDIR,
		    config => $Lab::TESTFILE,
		    filt => 1,
		    binsize => 10,
		    dev => '/xs',
		    );
my %opts = %default_opts;
GetOptions(\%opts,
	   'help!', 'version!', 'debug!',
	   'evtdir=s', 'bgdir=s', 'config=s', 'filt!', 'binsize=f', 'dev=s',
	   ) or die "Try --help for more information.\n";
if ($opts{debug}) {
  $SIG{__WARN__} = \&Carp::cluck;
  $SIG{__DIE__} = \&Carp::confess;
}
$opts{help} and _help();
$opts{version} and _version();

@ARGV and die "usage: $0 [options]\n";

$Lab::TESTFILE = $opts{config};

my (undef, undef, $mcp, $evt_time, $hrc_file, $bg_time, $bg_file)
   = Lab::test_data() or die "problem reading $opts{config}";

# pdls of X,Y event positions, exposure times for each chip_id
my (@rawx, @rawy, @exptime);

# initialze our data
for (0..2) {
  $rawx[$_] = long [];
  $rawy[$_] = $rawx[$_]->copy;
  $exptime[$_] = 0;
}

my ($rawx, $rawy, $chip_id);

my $suffix = '_evt1.fits';
$suffix = '_evt1_filt.fits' if $opts{filt};

for my $i (0..$#{$hrc_file}) {

  my $mcp = $mcp->[$i];
  my $hrc_file = $hrc_file->[$i];
  my $bg_file = $bg_file->[$i];
  my $bg_time = $bg_time->[$i];
  my $evt_time = $evt_time->[$i];

  my $evtfile = $opts{evtdir}. '/' . $hrc_file . $suffix;
  my $bgfile = $opts{bgdir}. '/' . $bg_file . $suffix;

  # read bg data
  print STDERR "reading $bgfile...";
  STDERR->flush;
  ($rawx, $rawy, $chip_id) = read_bintbl_cols($bgfile, 'rawx', 'rawy', 'chip_id', { status => 1, extname => 'events' }) or die;
  print STDERR " done\n";

  for (0..2) {
    my $ind = which($chip_id == $_+1);
    $rawx[$_] = $rawx[$_]->append($rawx->index($ind));
    $rawy[$_] = $rawy[$_]->append($rawy->index($ind));
    $exptime[$_] += $bg_time;
  }

  # do same for evt file, skipping current active MCP
  print STDERR "reading $evtfile...";
  STDERR->flush;
  ($rawx, $rawy, $chip_id) = read_bintbl_cols($evtfile, 'rawx', 'rawy', 'chip_id', { status => 1, extname => 'events' }) or die;
  print STDERR " done\n";

  my $chip_id_active = Lab::mcp_to_chipid($mcp);;
  for (0..2) {
    next if $_+1 == $chip_id_active;
    my $ind = which($chip_id == $_+1);
    $rawx[$_] = $rawx[$_]->append($rawx->index($ind));
    $rawy[$_] = $rawy[$_]->append($rawy->index($ind));
    $exptime[$_] += $evt_time;
  }

}

dev $opts{dev}, 2, 3;

for (0..2) {
  my $rawx = $rawx[$_];
  my $rawy = $rawx[$_];
  my $exptime = $exptime[$_];
  my $chip_id = $_ + 1;

  my ($xmin, $xmax) = $rawx->minmax;
  my ($ymin, $ymax) = $rawy->minmax;

  my ($histx, $histy);

  # rawx histogram, counts / sec / pixel
  ($histx, $histy) = hist($rawx, $xmin, $xmax, $opts{binsize});
  $histy = float($histy) / $opts{binsize} / ($ymax-$ymin) / $opts{binsize} / $exptime;
  bin $histx, $histy, { xtitle => 'rawx', ytitle => 'cnts / sec / pixel\\u2\\d', title => "chip_id $chip_id" };

  # rawy histogram, counts / sec / pixel
  ($histx, $histy) = hist($rawy, $ymin, $ymax, $opts{binsize});
  $histy = float($histy) / $opts{binsize} / ($xmax-$xmin) / $opts{binsize} / $exptime;
  bin $histx, $histy, { xtitle => 'rawy', ytitle => 'cnts / sec / pixel\\u2\\d', title => "chip_id $chip_id" };

}

exit 0;

sub _help {
  exec("$Config{installbin}/perldoc", '-F', $FindBin::Bin . '/' . $FindBin::RealScript);
}

sub _version {
  print $version,"\n";
  exit 0;
}

File: bg_rates.cc
#include <vector>
#include <string>
#include <cstdlib>
#include <iterator>
#include <cstdio>
#include <getopt.h>
#include <iostream>
#include "lab.hh"

using std::cout;
using std::cerr;
using std::vector;
using std::string;
using std::ostream_iterator;

namespace {

  namespace opts {
    const char* config = lab::testfile;
    const char* evtdir = lab::evtdir;
    const char* bgdir = lab::bgdir;
    int filter = 1;
  
    char* version_string = "0.1";
    int help = 0;
    int version = 0;
    option lopts[] = {
      { "help",    no_argument, &help, 1 },
      { "version", no_argument, &version, 1 },
      { "nofilter", no_argument, &filter, 0 },
      { 0, 0, 0, 0 }
    };
  }

  int help();
  int version();
}

int main(int argc, char** argv) {

  int c;
  while ((c=getopt_long_only(argc, argv, "", opts::lopts, 0))!=-1) {
    switch (c) {
    // a flag was set/unset on our behalf, nothing more to do
    case 0:
      break;
    // problem occurred
    case '?':
    case ':':
      cerr << "Try `--help' for more information.\n";
      return EXIT_FAILURE;
    // didn't handle all of our specified options
    default:
      cerr << "programmer error, unhandled option = "; cerr.put(c); cerr << '\n';
      return EXIT_FAILURE;
    }
  }

  if (opts::help) return help();
  if (opts::version) return version();

  if ( argc-optind != 0) {
    cerr << "Usage: " << argv[0] << " [options] file\n";
    return EXIT_FAILURE;
  }

  vector<string> line;
  vector<int> energy;
  vector<int> mcp;
  vector<int> time;
  vector<int> bg_time;
  vector<string> hrc_file;
  vector<string> bg_hrc_file;

  lab::test_data(vector<string>(), line, energy, mcp, time, hrc_file, bg_time, bg_hrc_file);

  copy(line.begin(), line.end(), ostream_iterator<string>(cout, " "));
  cout << '\n';

  copy(hrc_file.begin(), hrc_file.end(), ostream_iterator<string>(cout, " "));
  cout << '\n';

  copy(bg_hrc_file.begin(), bg_hrc_file.end(), ostream_iterator<string>(cout, " "));
  cout << '\n';

  copy(mcp.begin(), mcp.end(), ostream_iterator<int>(cout, " "));
  cout << '\n';

  copy(energy.begin(), energy.end(), ostream_iterator<int>(cout, " "));
  cout << '\n';

  copy(time.begin(), time.end(), ostream_iterator<int>(cout, " "));
  cout << '\n';

  copy(bg_time.begin(), bg_time.end(), ostream_iterator<int>(cout, " "));
  cout << '\n';

} // main

namespace {

  int version() {
    cout << opts::version_string << '\n';
    return 0;
  }

  int help() {
    const char* help_text = "\
=head1 NAME\n\
\n\
template - a template C++ program\n\
\n\
=head1 SYNOPSIS\n\
\n\
template [options] arg1 arg2\n\
\n\
=head1 DESCRIPTION\n\
\n\
A template for a C++ program with option processing and --help\n\
text placeholder.\n\
\n\
=head1 OPTIONS\n\
\n\
=over 4\n\
\n\
=item --help\n\
\n\
Print this help text and exit.\n\
\n\
=item --version\n\
\n\
Print the program version and exit.\n\
\n\
=back\n\
\n\
=head1 AUTHOR\n\
\n\
Peter Ratzlaff E<lt>pratzlaff@cfa.harvard.eduE<gt> August 2007\n\
\n\
=head1 SEE ALSO\n\
\n\
nothing to see here\n\
\n\
=cut\n\
";

    const char* pager = std::getenv("PAGER");
    if (!pager) pager = "more";

    FILE* pd = popen((std::string("pod2text -c | ")+pager).c_str(), "w");
    if (!pd) {
      std::perror("error starting pod2text");
      return EXIT_FAILURE;
    }

    int n = 0;
    int len = std::strlen(help_text);
    while (n < len) {
      int written = std::fwrite(help_text, 1, len-n, pd);
      if (!written) {
	std::perror("error writing help");
	return EXIT_FAILURE;
      }
      n+=written;
    }

    if (pclose(pd) == -1) {
      std::perror("error writing help");
      return EXIT_FAILURE;
    }

    return 0;
  }

}
